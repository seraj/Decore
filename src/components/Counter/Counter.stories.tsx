import { text, number, boolean, withKnobs } from "@storybook/addon-knobs";
import { storiesOf } from "@storybook/react";
import * as React from "react";
import wInfo from "../../utils/wInfo";
import Counter from "./Counter";

const stories = storiesOf("Components/Counter", module);
stories.addDecorator(withKnobs);

stories.add(
  "default",
  wInfo()(() => (
    <Counter
      min={number("min", -10)}
      max={number("max", 10)}
      defaultValue={text("defaultValue", "0")}
      value={number("value", undefined)}
      disabled={boolean("disabled", false)}
    ></Counter>
  ))
);
