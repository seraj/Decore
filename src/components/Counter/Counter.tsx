import * as React from "react";
import StyledCounter, { Container, HiddenInput } from "./styled/StyledCounter";
import CounterProps from "./Counter.props";

interface CounterState {
  value: number | undefined;
  userSet: boolean;
}

class Counter extends React.Component<CounterProps, CounterState> {
  public state = {
    value: 0,
    userSet: false
  };

  public static getDerivedStateFromProps(
    props: CounterProps,
    state: CounterState
  ) {
    const propsValue = Counter.parseValue(
      props.value,
      Counter.parseValue(props.defaultValue, 0)
    );
    return {
      value: state.userSet ? state.value : propsValue
    };
  }

  private boundCheck = (v: number) => {
    let { min, max } = this.props;
    if (!min) {
      min = -Infinity;
    } else if (typeof min == "string") {
      min = parseInt(min, 10);
    }

    if (!max) {
      max = Infinity;
    } else if (typeof max == "string") {
      max = parseInt(max, 10);
    }
    return Math.max(Math.min(v, max), min);
  };

  private increment = () => {
    this.setState({
      value: this.boundCheck(this.state.value + 1),
      userSet: true
    });
  };

  private decrement = () => {
    this.setState({
      value: this.boundCheck(this.state.value - 1),
      userSet: true
    });
  };

  public static parseValue(
    v: string | string[] | undefined | number,
    defaultValue: number | undefined
  ) {
    if (v === undefined) {
      return defaultValue;
    } else if (typeof v == "string") {
      return parseInt(v, 10);
    } else if (Array.isArray(v)) {
      return parseInt(v[0], 10);
    }

    return v;
  }

  public render() {
    const { children, min, max, ...props } = this.props;

    let value: number | undefined = this.state.value;
    if (this.props.value !== undefined) {
      value = Counter.parseValue(this.props.value, undefined);
    }

    return (
      <Container>
        <HiddenInput value={value} {...props} type="number" />
        <StyledCounter {...props} value={value} min={min} max={max}>
          <span className="decore__minus" onClick={this.decrement}>
            -
          </span>
          <span className="decore__value">{value}</span>
          <span className="decore__plus" onClick={this.increment}>
            +
          </span>
        </StyledCounter>
      </Container>
    );
  }
}

export default Counter;
