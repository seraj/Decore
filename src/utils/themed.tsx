import * as React from "react";
import * as renderer from "react-test-renderer";
import Decore from "../components/Decore";
import theme from "../theme";

const themed = (component: JSX.Element) =>
  renderer.create(<Decore theme={theme}>{component}</Decore>);

export default themed;
