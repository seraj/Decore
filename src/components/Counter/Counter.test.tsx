import * as React from "react";

import themed from "../../utils/themed";
import Counter from "./Counter";

describe("Counter", () => {
  it("should render correctly", () => {
    const component = themed(<Counter>Test</Counter>).toJSON();

    expect(component).toMatchSnapshot();
  });
});
