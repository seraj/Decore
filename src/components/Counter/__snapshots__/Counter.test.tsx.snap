// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should render correctly 1`] = `
.c2 {
  border: 1.5px solid #999999;
  background-color: undefined;
  padding: 1rem 0.8rem;
  height: 45px;
  border-radius: 0px;
  color: #333333;
}

.c2:hover {
  border-color: #333333;
}

.c2:disabled {
  background: #E5E5E5;
  color: #999999;
  cursor: not-allowed !important;
  pointer-events: none;
}

.c0 {
  display: inline-block;
  vertical-align: middle;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px !important;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px !important;
}

.c3 {
  display: inline-block;
  width: 78px;
  height: 29px;
  background: #ffffff;
  border-color: #999999;
  border-width: 1.5px;
  border-style: solid;
  position: relative;
}

.c3:hover {
  border-color: #333333;
}

.c3 .decore__value {
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  line-height: 29px;
  color: #999999;
  text-align: center;
  display: block;
  font-size: 0.8em;
}

.c3 .decore__minus,
.c3 .decore__plus {
  position: absolute;
  width: 25px;
  text-align: center;
  top: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #333333;
  cursor: pointer;
  z-index: 1;
}

.c3 .decore__minus {
  left: 0;
  line-height: 26px;
}

.c3 .decore__minus:hover {
  background: #E5E5E5;
}

.c3 .decore__plus {
  right: 0;
  line-height: 26px;
}

.c3 .decore__plus:hover {
  background: #E5E5E5;
}

<div
  className="c0"
>
  <input
    className="c1 c2"
    type="number"
    value={0}
  />
  <div
    className="c3"
    value={0}
  >
    <span
      className="decore__minus"
      onClick={[Function]}
    >
      -
    </span>
    <span
      className="decore__value"
    >
      0
    </span>
    <span
      className="decore__plus"
      onClick={[Function]}
    >
      +
    </span>
  </div>
</div>
`;
